server/pkg/nlslayer/nlslayer_list.go

```diff
@@ -0,0 +1,296 @@
+	res := make(NLSLayerSimpleList, 0, len(ll))
+	for _, l := range ll {
+		if li := NLSLayerSimpleFromLayerRef(l); li != nil {
+			res = append(res, li)
+		}
+	}
+	return res
+}
+
+func (ll NLSLayerList) ToNLSLayerGroupList() NLSLayerGroupList {
+	res := make(NLSLayerGroupList, 0, len(ll))
+	for _, l := range ll {
+		if lg := NLSLayerGroupFromLayerRef(l); lg != nil {
+			res = append(res, lg)
+		}
+	}
+	return res
+}
+
+func (ll NLSLayerList) SeparateLayerItemAndGroup() (NLSLayerSimpleList, NLSLayerGroupList) {
+	resi := make(NLSLayerSimpleList, 0, len(ll))
+	resg := make(NLSLayerGroupList, 0, len(ll))
+	for _, l := range ll {
+		if lg := NLSLayerGroupFromLayerRef(l); lg != nil {
+			resg = append(resg, lg)
+		} else if li := NLSLayerSimpleFromLayerRef(l); li != nil {
+			resi = append(resi, li)
+		}
+	}
+	return resi, resg
+}
+
+func (ll NLSLayerList) Deref() []NLSLayer {
+	if ll == nil {
+		return nil
+	}
+	res := make([]NLSLayer, 0, len(ll))
+	for _, l := range ll {
+		if l != nil {
+			res = append(res, *l)
+		} else {
+			res = append(res, nil)
+		}
+	}
+	return res
+}
+
+func (ll NLSLayerList) Loader() Loader {
+	return LoaderFrom(ll.Deref())
+}
+
+func (ll NLSLayerList) Map() Map {
+	m := make(Map, len(ll))
+	m.Add(ll...)
+	return m
+}
+
+func (ll NLSLayerList) Remove(lids ...ID) NLSLayerList {
+	if ll == nil {
+		return nil
+	}
+
+	res := make(NLSLayerList, 0, len(ll))
+
+	for _, l := range ll {
+		if l == nil {
+			continue
+		}
+		hit := false
+		for _, lid := range lids {
+			if (*l).ID() == lid {
+				hit = true
+				break
+			}
+		}
+		if !hit {
+			res = append(res, l)
+		}
+	}
+
+	return res
+}
+
+func (ll NLSLayerList) AddUnique(newList ...*NLSLayer) NLSLayerList {
+	res := append(NLSLayerList{}, ll...)
+
+	for _, l := range newList {
+		if l == nil {
+			continue
+		}
+		if res.Find((*l).ID()) != nil {
+			continue
+		}
+		res = append(res, l)
+	}
+
+	return res
+}
+
+type NLSLayerSimpleList []*NLSLayerSimple
+
+
+func (ll NLSLayerSimpleList) ToLayerList() NLSLayerList {
+	res := make(NLSLayerList, 0, len(ll))
+	for _, l := range ll {
+		var layer NLSLayer = l
+		res = append(res, &layer)
+	}
+	return res
+}
+
+func (ll NLSLayerSimpleList) Last() *NLSLayerSimple {
+	if len(ll) == 0 {
+		return nil
+	}
+	return ll[len(ll)-1]
+}
+
+type NLSLayerGroupList []*NLSLayerGroup
+
+func (ll NLSLayerGroupList) ToLayerList() NLSLayerList {
+	res := make(NLSLayerList, 0, len(ll))
+	for _, l := range ll {
+		var layer NLSLayer = l
+		res = append(res, &layer)
+	}
+	return res
+}
+
+func (ll NLSLayerGroupList) Last() *NLSLayerGroup {
+	if len(ll) == 0 {
+		return nil
+	}
+	return ll[len(ll)-1]
+}
+
+type Map map[ID]*NLSLayer
+
+func MapFrom(l NLSLayer) Map {
+	return NLSLayerList{&l}.Map()
+}
+
+func (m Map) Add(layers ...*NLSLayer) Map {
+	if m == nil {
+		m = map[ID]*NLSLayer{}
+	}
+	for _, l := range layers {
+		if l == nil {
+			continue
+		}
+		l2 := *l
+		if l2 == nil {
+			continue
+		}
+		m[l2.ID()] = l
```

Comment:
I did not get the idea here!

Reply:
1. If the map (m) is nil, it initializes it to an empty map of the right type.
2. It iterates over the provided layers. For each layer:
  - If the layer pointer (l) is nil, it skips to the next iteration.
  - Dereferences the layer pointer to get the layer object (l2).
  - If l2 is nil (which means that the pointer points to a nil value), it skips to the next iteration.
  - Uses the ID() method of the layer (l2.ID()) to fetch its unique ID and adds the layer pointer to the map with that ID as the key.
3. Finally, it returns the modified (or new) map.